### PR TITLE
Provision: Replace virtualenv by python native venv.

### LINKED ISSUE #625: Move away from virtualenv
Since python3.5, the recommended methodology for creating virtual environments is to use the native venv module.

Now that we no longer support older versions of python, we no longer need virtualenv.

More discussion here: https://chat.zulip.org/#narrow/stream/2-general/topic/page.20not.20found/near/1022651

### PR DESCRIPTION
**Pull request by LoopThrough-i-j**
Fixes #625

### CODE REVIEW

**Review by LoopThrough-i-j — COMMENTED** [review]
I am not sure about this line. Can anyone help?

File: tools/provision [hunk 1]
```diff
@@ -39,7 +37,7 @@ the Python version this command is executed with."""
     py_version = tuple(int(num) for num in py_version_list[0:2])
     venv_name = 'zulip-api-py{}-venv'.format(py_version[0])
 
-    if py_version <= (3, 1) and (not options.force):
+    if py_version <= (3, 5) and (not options.force):
```
Comment (rht) [inline tools/provision:40]: This version constraint update should be in a separate commit.

File: tools/provision [hunk 2]
```diff
@@ -48,37 +46,31 @@ the Python version this command is executed with."""
...
     if not os.path.isdir(venv_dir):
         try:
-            return_code = subprocess.call(['virtualenv', '-p', options.python_interpreter, venv_dir])
+            return_code = subprocess.call([options.python_interpreter, '-m', 'venv', venv_dir])
         except OSError:
-            if subprocess.call(['which', 'virtualenv']):
-                print("{red}Please install the virtualenv package and try again.{end_format}"
-                      .format(red='\033[91m', end_format='\033[0m'))
-                sys.exit(1)
+            print("{red}Installation with venv failed."
```
Comment (rht) [inline tools/provision:51]: I think on Ubuntu, `python3-venv` is a separate package (but not sure, haven't checked). If so, this err message doesn't take into account of the fact that the library hasn't been installed yet.

File: tools/provision [hunk 3]
```diff
@@ -48,37 +46,31 @@ the Python version this command is executed with."""
...
     else:
         venv_exec_dir = 'bin'
 
-    # In order to install all required packages for the venv, we need to activate it. Since
-    # the activation script sets environmental variables, it needs to be executed inline with
-    # `import_module`.
-    activate_module_dir = os.path.abspath(os.path.join(venv_dir, venv_exec_dir))
-    sys.path.append(activate_module_dir)
-
-    import_module('activate_this')
```
Comment (rht) [inline tools/provision:80]: Why is this no longer necessary?

File: tools/provision (hunk 2 above)
Comment (rht) [reply tools/provision:51]: Yes, confirmed that it is a separate package: https://packages.ubuntu.com/search?keywords=python3-venv. Note that this is specific to Ubuntu only.

File: tools/provision (hunk 3 above)
Comment (LoopThrough-i-j) [reply tools/provision:80]: 1. There is no activate_this in venv 
2. If you check the way packages are getting installed we don't actually need to activate the environment. To install packages we are fetching the path to pip in the venv and installing packages using that, which limits the package installation to the environment.

File: tools/provision (hunk 2 above)
Comment (LoopThrough-i-j) [reply tools/provision:51]: Yes, I know that as well. So do I mention that in the error message or run a subprocess for that? I think we can just add an extra line for that.

File: tools/provision (hunk 1 above)
Comment (LoopThrough-i-j) [reply tools/provision:40]: okay, will do sure.

File: tools/provision (hunk 2 above)
Comment (rht) [reply tools/provision:51]: Yes, you should say in error message that if you are on Ubuntu, you haven't installed `python3-venv`.

File: tools/provision (hunk 1 above)
Comment (rht) [inline tools/provision:40]: I grepped for 3.5 in the codebase. It's not only in this line that the version needs to be updated. @LoopThrough-i-j I think you should do the version constraint thing in a separate PR.

e.g.
```
zulip_botserver/setup.py
27:        'Programming Language :: Python :: 3.5',
32:    python_requires='>=3.5',
```

and several more.

Comment (LoopThrough-i-j) [reply tools/provision:40]: ok makes sense will do.

**Comment by timabbott** [comment]
Merged, thanks @LoopThrough-i-j!  I made a few edits to the commit message, mainly for technical English ("Replace virtualenv with python native venv.", i.e. I did s/by/with/).

File: tools/provision (hunk 1 above)
Comment (rht) [reply tools/provision]: @LoopThrough-i-j have you made the PR to drop 3.5 yet?

Comment (rht) [reply tools/provision]: #655

**Comment by LoopThrough-i-j** [comment]
Thanks, will keep it in mind.
